addons/bus/static/src/websocket_service.js

```diff
@@ -0,0 +1,92 @@
+/** @odoo-module **/
+import { browser } from "@web/core/browser/browser";
+import { registry } from '@web/core/registry';
+
+import { WebsocketError } from './websocket_errors';
+
+const { EventBus } = owl;
+/**
+ * Communicate with a SharedWorker in order to provide a single websocket
+ * connection shared across multiple tabs.
+ *
+ *  @emits connect
+ *  @emits disconnect
+ *  @emits reconnect
+ *  @emits reconnecting
+ *  @emits notification
+ */
+export const websocketService = {
+    dependencies: ['notification', 'user'],
+    _removeConnectionLostNotification: null,
+
+    start(env) {
+        const bus = new EventBus();
```

Comment:
I'm really not fond of `EventBus` all over the place. It makes things very hard to trace because lots of jumps in the code, and also the API are not properly named (i.e. not proper method names, they are data payload, so hard to see the actual flow just at a glance at at the code)

The whole code would be far far easier to understand if instead you implement in a declarative way.
I would highly suggest making models for the code of these services, as it would make things much easier to understand, and so make code much more maintainable